api/authentication/authentication_test.go

```diff
@@ -0,0 +1,173 @@
+func newTestClient(addr string, credFn func() (credentials.TransportCredentials, error)) (*grpc.ClientConn, error) {
+	creds, err := credFn()
+	if err != nil {
+		return nil, err
+	}
+
+	return grpc.Dial(addr, grpc.WithTransportCredentials(creds))
+}
+
+func knownUserCredentials() (credentials.TransportCredentials, error) {
+	return NewClientTransportCredentials(
+		"test-assets/user.key",
+		"test-assets/user.crt",
+		"test-assets/ca.crt",
+	)
+}
+
+// unknownUserCredentials returns user credentials with a certificate signed by
+// an unknown CA
+func unknownUserCredentials() (credentials.TransportCredentials, error) {
+	return NewClientTransportCredentials(
+		"test-assets/unknown-user.key",
+		"test-assets/unknown-user.crt",
+		"test-assets/ca.crt",
+	)
+}
+
+func TestUnknownServerAuthentication(t *testing.T) {
+	srv, l, err := newTestServer(unknownServerCredentials)
+	assertNil(t, err)
+
+	go srv.Serve(l)
+	defer srv.Stop()
+
+	conn, err := newTestClient(getServerAddress(l), knownUserCredentials)
+	assertNil(t, err)
+	defer conn.Close()
+
+	err = conn.Invoke(context.Background(), "fake", nil, nil)
+	assertNonNil(t, err)
+
+	statusCode := status.Convert(err).Code()
+	if statusCode != codes.Unavailable {
+		t.Fatalf("'%s' expected, '%s' got", codes.Unavailable, statusCode)
+	}
+}
+
+func TestKnownClientAuthentication(t *testing.T) {
+	srv, l, err := newTestServer(knownServerCredentials)
+	assertNil(t, err)
+
+	go srv.Serve(l)
+	defer srv.Stop()
+
+	conn, err := newTestClient(getServerAddress(l), knownUserCredentials)
+	assertNil(t, err)
+	defer conn.Close()
+
+	err = conn.Invoke(context.Background(), "fake", nil, nil)
+	assertNonNil(t, err)
+
+	statusCode := status.Convert(err).Code()
+	if statusCode != codes.Unimplemented {
+		t.Fatalf("'%s' expected, '%s' got", codes.Unimplemented, statusCode)
+	}
+}
+
+// TestUnknownClientAuthentication tests a user certificate signed by an unknown
+// CA
+func TestUnknownClientAuthentication(t *testing.T) {
+	srv, l, err := newTestServer(knownServerCredentials)
+	assertNil(t, err)
+
+	go srv.Serve(l)
+	defer srv.Stop()
+
+	conn, err := newTestClient(getServerAddress(l), unknownUserCredentials)
+	assertNil(t, err)
+	defer conn.Close()
+
+	err = conn.Invoke(context.Background(), "fake", nil, nil)
+	assertNonNil(t, err)
+
+	statusCode := status.Convert(err).Code()
+	if statusCode != codes.Unavailable {
+		t.Fatalf("'%s' expected, '%s' got", codes.Unavailable, statusCode)
+	}
+}
+
+func TestInvalidCreds(t *testing.T) {
+	cs := [][]string{
+		// Invalid key
+		{"authentication_test.go", "test-assets/server.crt", "test-assets/ca.crt"},
+		{"/", "test-assets/server.crt", "test-assets/ca.crt"},
+
+		// Invalid cert
+		{"test-assets/server.key", "authentication_test.go", "test-assets/ca.crt"},
+		{"test-assets/server.key", "/", "test-assets/ca.crt"},
+
+		// Invalid CA
+		{"test-assets/server.key", "test-assets/server.crt", "authentication_test.go"},
+		{"test-assets/server.key", "test-assets/server.crt", "/"},
+	}
```

Comment:
It hard to follow where args goes and in case of error it will be hard to debug with case failed. Test table should improve that:

```suggestion
	cs := []struct{
		desc string
		keyFile string
		certFile string
		caFile string
	} {
	    {
	        desc: "Invalid Key",
	        ...
	     }, 
	     {
	        ...
	     }
	}
	for _, tc := range cs {
		t.Run(tc.desc, func(t *testing.T){ 
		        ...
		}
	}
```